<<<<<<< HEAD
# Disallow explicitly returning from tests (`no-test-return-statement`)

<!-- end auto-generated rule header -->
=======
# Disallow explicitly returning from tests (no-test-return-statement)
>>>>>>> 086bb10 (Add utils and controllers updates)

Tests in Jest should be void and not return values.

If you are returning Promises then you should update the test to use
`async/await`.

## Rule details

<<<<<<< HEAD
This rule triggers a warning if you use a return statement inside a test body.
=======
This rule triggers a warning if you use a return statement inside of a test
body.
>>>>>>> 086bb10 (Add utils and controllers updates)

```js
/*eslint jest/no-test-return-statement: "error"*/

// valid:

<<<<<<< HEAD
it('noop', function () {});
=======
it('noop', function() {});
>>>>>>> 086bb10 (Add utils and controllers updates)

test('noop', () => {});

test('one arrow', () => expect(1).toBe(1));

test('empty');

test('one', () => {
  expect(1).toBe(1);
});

<<<<<<< HEAD
it('one', function () {
=======
it('one', function() {
>>>>>>> 086bb10 (Add utils and controllers updates)
  expect(1).toBe(1);
});

it('returning a promise', async () => {
  await new Promise(res => setTimeout(res, 100));
  expect(1).toBe(1);
});

// invalid:
test('return an expect', () => {
  return expect(1).toBe(1);
});

<<<<<<< HEAD
it('returning a promise', function () {
=======
it('returning a promise', function() {
>>>>>>> 086bb10 (Add utils and controllers updates)
  return new Promise(res => setTimeout(res, 100)).then(() => expect(1).toBe(1));
});
```
